File: backend/src/engine/sts.py
"""Short-Term Scheduler (STS) — MLFQ Priority Queues.

Modified Multilevel Feedback Queue adapted for human tasks:
  P0 (Urgent):     Hard deadlines within 2 hours, external dependencies
  P1 (Important):  Deadlines today, high-impact, upstream blockers
  P2 (Normal):     Routine tasks, flexible deadlines, personal goals
  P3 (Background): Nice-to-haves, low-energy fillers, delegatable

Energy constraint: never schedule energy_cost > energy_level.
"""

from __future__ import annotations

import heapq
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Optional

from src.models.task import Task, Priority, TaskStatus


@dataclass(order=True)
class _QueueEntry:
    """Priority queue entry. Lower sort_key = higher priority."""
    sort_key: float
    task: Task = field(compare=False)


class ShortTermScheduler:
    """MLFQ with 4 priority levels and energy-aware scheduling."""

    def __init__(self):
        # 4 priority queues (min-heaps)
        self._queues: dict[int, list[_QueueEntry]] = {
            Priority.P0_URGENT: [],
            Priority.P1_IMPORTANT: [],
            Priority.P2_NORMAL: [],
            Priority.P3_BACKGROUND: [],
        }
        # Currently executing task (for preemption)
        self._current_task: Optional[Task] = None
        # Tasks delegated to GhostWorker
        self._delegation_queue: list[Task] = []

    def enqueue(self, task: Task) -> None:
        """Add a task to the appropriate priority queue."""
        priority = self._classify_priority(task)
        task.priority = priority
        # Sort within priority level by deadline urgency (higher urgency = lower sort_key)
        sort_key = -task.deadline_urgency
        heapq.heappush(self._queues[priority], _QueueEntry(sort_key, task))

    def enqueue_batch(self, tasks: list[Task]) -> None:
        """Add multiple tasks, then rebalance."""
        for task in tasks:
            self.enqueue(task)

    def dequeue(self, energy_level: int = 5) -> Optional[Task]:
        """Get the next task to execute, respecting energy constraints.

        Scans from P0 to P3. Skips tasks whose energy_cost > energy_level.
        If energy is low and only P3 tasks remain, auto-delegates them.
        """
        for priority in (Priority.P0_URGENT, Priority.P1_IMPORTANT,
                         Priority.P2_NORMAL, Priority.P3_BACKGROUND):
            queue = self._queues[priority]
            # Find first task that fits energy budget
            skipped = []
            result = None
            while queue:
                entry = heapq.heappop(queue)
                if entry.task.energy_cost <= energy_level:
                    result = entry.task
                    # Put skipped entries back
                    for s in skipped:
                        heapq.heappush(queue, s)
                    break
                skipped.append(entry)
            else:
                # Put all skipped entries back if nothing found
                for s in skipped:
                    heapq.heappush(queue, s)

            if result:
                return result

        return None

    def preempt(self, urgent_task: Task, energy_level: int = 5) -> Optional[Task]:
        """Preemptively interrupt current task for a more urgent one.

        Returns the preempted task (with saved state) or None if no preemption needed.
        """
        if not self._current_task:
            self._current_task = urgent_task
            return None

        current_priority = self._current_task.priority
        urgent_priority = self._classify_priority(urgent_task)

        if urgent_priority < current_priority:  # lower number = higher priority
            # Save state of current task
            preempted = self._current_task
            preempted.status = TaskStatus.ACTIVE  # back to ready queue
            # Re-enqueue preempted task
            self.enqueue(preempted)
            # Switch to urgent task
            self._current_task = urgent_task
            return preempted

        # Not urgent enough to preempt; just enqueue
        self.enqueue(urgent_task)
        return None

    def set_current(self, task: Task) -> None:
        """Mark a task as currently executing."""
        self._current_task = task
        task.status = TaskStatus.IN_PROGRESS

    def get_current(self) -> Optional[Task]:
        return self._current_task

    def clear_current(self) -> None:
        self._current_task = None

    def auto_delegate_p3(self, energy_level: int) -> list[Task]:
        """When energy is low, delegate all P3 tasks to GhostWorker."""
        if energy_level > 2:
            return []

        delegated = []
        queue = self._queues[Priority.P3_BACKGROUND]
        while queue:
            entry = heapq.heappop(queue)
            entry.task.status = TaskStatus.DELEGATED
            delegated.append(entry.task)
            self._delegation_queue.append(entry.task)
        return delegated

    def get_delegation_queue(self) -> list[Task]:
        """Get and clear the delegation queue."""
        q = list(self._delegation_queue)
        self._delegation_queue.clear()
        return q

    def get_ordered_schedule(self, energy_level: int = 5) -> list[Task]:
        """Return all queued tasks in execution order (non-destructive).

        Respects energy constraints: tasks above energy budget are placed
        at the end.
        """
        schedule = []
        deferred = []

        for priority in (Priority.P0_URGENT, Priority.P1_IMPORTANT,
                         Priority.P2_NORMAL, Priority.P3_BACKGROUND):
            entries = sorted(self._queues[priority])
            for entry in entries:
                if entry.task.energy_cost <= energy_level:
                    schedule.append(entry.task)
                else:
                    deferred.append(entry.task)

        return schedule + deferred

    def reorder(self, tasks: list[Task]) -> None:
        """Clear and rebuild all queues from a list of tasks."""
        for q in self._queues.values():
            q.clear()
        self.enqueue_batch(tasks)

    def _classify_priority(self, task: Task) -> int:
        """Auto-classify task priority based on deadline and attributes."""
        # If task already has a manually set priority, respect it
        if task.priority != Priority.P2_NORMAL:
            return task.priority

        now = datetime.now(timezone.utc)

        if task.deadline:
            try:
                dl = datetime.fromisoformat(task.deadline)
                hours_left = (dl - now).total_seconds() / 3600
                if hours_left <= 2:
                    return Priority.P0_URGENT
                elif hours_left <= 24:
                    return Priority.P1_IMPORTANT
            except (ValueError, TypeError):
                pass

        if task.cognitive_load <= 1 and task.energy_cost <= 1:
            return Priority.P3_BACKGROUND

        return task.priority

    @property
    def total_count(self) -> int:
        return sum(len(q) for q in self._queues.values())

    def queue_counts(self) -> dict[str, int]:
        return {
            "P0_URGENT": len(self._queues[Priority.P0_URGENT]),
            "P1_IMPORTANT": len(self._queues[Priority.P1_IMPORTANT]),
            "P2_NORMAL": len(self._queues[Priority.P2_NORMAL]),
            "P3_BACKGROUND": len(self._queues[Priority.P3_BACKGROUND]),
        }

File: rewind-core/src/sts.rs
//! Short-Term Scheduler (STS) — modified MLFQ priority queues.
//!
//! Port target: `backend/src/engine/sts.py` (origin/main).

use crate::task::{Priority, Task, TaskStatus};
use chrono::{DateTime, Utc};
use std::cmp::Ordering;
use std::collections::BinaryHeap;

#[derive(Debug, Clone)]
struct QueueEntry {
    // Lower sort_key = higher priority within a given priority queue.
    // We use -deadline_urgency so bigger urgency comes first.
    sort_key: i32,
    seq: u64,
    task: Task,
}

impl PartialEq for QueueEntry {
    fn eq(&self, other: &Self) -> bool {
        self.sort_key == other.sort_key && self.seq == other.seq
    }
}
impl Eq for QueueEntry {}

impl PartialOrd for QueueEntry {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for QueueEntry {
    fn cmp(&self, other: &Self) -> Ordering {
        // BinaryHeap is max-heap.
        // We want the smallest sort_key first, then smallest seq.
        // So we invert ordering.
        other
            .sort_key
            .cmp(&self.sort_key)
            .then_with(|| other.seq.cmp(&self.seq))
    }
}

#[derive(Debug, Default)]
pub struct ShortTermScheduler {
    queues: [BinaryHeap<QueueEntry>; 4],
    current_task: Option<Task>,
    delegation_queue: Vec<Task>,
    seq: u64,
}

impl ShortTermScheduler {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn enqueue(&mut self, mut task: Task, now: DateTime<Utc>) {
        let p = self.classify_priority(&task, now);
        task.priority = p;

        let idx = priority_index(p);
        let sort_key = -task.deadline_urgency;

        self.seq += 1;
        self.queues[idx].push(QueueEntry {
            sort_key,
            seq: self.seq,
            task,
        });
    }

    pub fn enqueue_batch(&mut self, tasks: Vec<Task>, now: DateTime<Utc>) {
        for t in tasks {
            self.enqueue(t, now);
        }
    }

    /// Dequeue next runnable task respecting energy constraints.
    pub fn dequeue(&mut self, energy_level: i32) -> Option<Task> {
        for p in [Priority::P0Urgent, Priority::P1Important, Priority::P2Normal, Priority::P3Background] {
            let idx = priority_index(p);

            // Pop until we find an energy-compatible task, buffering skipped.
            let mut skipped: Vec<QueueEntry> = Vec::new();
            let mut result: Option<Task> = None;

            while let Some(entry) = self.queues[idx].pop() {
                if entry.task.energy_cost <= energy_level {
                    result = Some(entry.task);
                    break;
                }
                skipped.push(entry);
            }

            // Put skipped entries back.
            for e in skipped {
                self.queues[idx].push(e);
            }

            if result.is_some() {
                return result;
            }
        }

        None
    }

    pub fn set_current(&mut self, mut task: Task) {
        task.status = TaskStatus::InProgress;
        self.current_task = Some(task);
    }

    pub fn clear_current(&mut self) {
        self.current_task = None;
    }

    pub fn get_current(&self) -> Option<&Task> {
        self.current_task.as_ref()
    }

    /// Delegate all P3 tasks when energy is low.
    pub fn auto_delegate_p3(&mut self, energy_level: i32) -> Vec<Task> {
        if energy_level > 2 {
            return Vec::new();
        }

        let idx = priority_index(Priority::P3Background);
        let mut delegated = Vec::new();
        while let Some(mut entry) = self.queues[idx].pop() {
            entry.task.status = TaskStatus::Delegated;
            delegated.push(entry.task.clone());
            self.delegation_queue.push(entry.task);
        }
        delegated
    }

    pub fn delegation_queue(&mut self) -> Vec<Task> {
        let q = self.delegation_queue.clone();
        self.delegation_queue.clear();
        q
    }

    pub fn total_count(&self) -> usize {
        self.queues.iter().map(|q| q.len()).sum()
    }

    fn classify_priority(&self, task: &Task, now: DateTime<Utc>) -> Priority {
        // Respect manually set priority if not default.
        if task.priority != Priority::P2Normal {
            return task.priority;
        }

        if let Some(dl) = task.deadline {
            let secs_left = (dl - now).num_seconds();
            let hours_left = (secs_left as f64) / 3600.0;
            if hours_left <= 2.0 {
                return Priority::P0Urgent;
            }
            if hours_left <= 24.0 {
                return Priority::P1Important;
            }
        }

        if task.cognitive_load <= 1 && task.energy_cost <= 1 {
            return Priority::P3Background;
        }

        Priority::P2Normal
    }
}

fn priority_index(p: Priority) -> usize {
    match p {
        Priority::P0Urgent => 0,
        Priority::P1Important => 1,
        Priority::P2Normal => 2,
        Priority::P3Background => 3,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::task::Task;
    use chrono::Duration;

    #[test]
    fn test_priority_by_deadline_2h() {
        let now = Utc::now();
        let t = Task::new("t1", "due soon")
            .with_deadline(now + Duration::minutes(90))
            .with_deadline_urgency(10);

        let mut sts = ShortTermScheduler::new();
        sts.enqueue(t, now);
        assert_eq!(sts.total_count(), 1);

        let next = sts.dequeue(5).unwrap();
        assert_eq!(next.priority, Priority::P0Urgent);
    }

    #[test]
    fn test_priority_by_deadline_24h() {
        let now = Utc::now();
        let t = Task::new("t1", "due today")
            .with_deadline(now + Duration::hours(10))
            .with_deadline_urgency(7);

        let mut sts = ShortTermScheduler::new();
        sts.enqueue(t, now);
        let next = sts.dequeue(5).unwrap();
        assert_eq!(next.priority, Priority::P1Important);
    }

    #[test]
    fn test_energy_constraint_skips_high_energy() {
        let now = Utc::now();
        let high = Task::new("t1", "hard").with_energy(5).with_deadline_urgency(10);
        let low = Task::new("t2", "easy").with_energy(1).with_deadline_urgency(1);

        let mut sts = ShortTermScheduler::new();
        sts.enqueue(high, now);
        sts.enqueue(low, now);

        // With energy=1, should skip high-energy task.
        let next = sts.dequeue(1).unwrap();
        assert_eq!(next.id, "t2");
    }

    #[test]
    fn test_auto_delegate_p3() {
        let now = Utc::now();
        let bg = Task::new("t3", "low effort")
            .with_energy(1)
            .with_cognitive(1)
            .with_deadline_urgency(0);

        let mut sts = ShortTermScheduler::new();
        sts.enqueue(bg, now);

        let delegated = sts.auto_delegate_p3(2);
        assert_eq!(delegated.len(), 1);
        assert_eq!(delegated[0].status, TaskStatus::Delegated);
        assert_eq!(sts.total_count(), 0);
    }
}
